server/bleep/src/indexes/doc.rs

```diff
@@ -0,0 +1,988 @@
+                    .expect("failed to get document by address");
+                SearchResult::from_tantivy_document(retrieved_doc, &self.section_schema)
+            })
+            .filter_map(|search_result| {
+                let term_distances =
+                    terms
+                        .iter()
+                        .fold(Vec::new(), |mut acc: Vec<(&&str, usize)>, x| {
+                            let next = if let Some((term, dist)) = acc.last() {
+                                (x, dist + term.len())
+                            } else {
+                                (x, 0)
+                            };
+                            acc.push(next);
+                            acc
+                        });
+
+                let dq = term_distances
+                    .windows(2)
+                    .map(|window| {
+                        let d1 = window[0].1;
+                        let d2 = window[1].1;
+                        d2.abs_diff(d1).pow(2)
+                    })
+                    .sum::<usize>();
+
+                let dt = |positions: &[Option<usize>], max: usize| {
+                    positions
+                        .windows(2)
+                        .map(|window| {
+                            let d1 = window[0].unwrap_or(max);
+                            let d2 = window[1].unwrap_or(max);
+                            d2.abs_diff(d1).pow(2)
+                        })
+                        .sum::<usize>()
+                };
+
+                let distance_penalty = |dq, dt, terms: &[&str]| {
+                    let total_len = terms.iter().map(|s| s.len()).sum::<usize>();
+                    if terms.is_empty() || dt == 0 {
+                        1.
+                    } else {
+                        (dt as f32 / dq as f32) * (total_len as f32)
+                    }
+                };
+
+                let (mut header_score, mut header_positions) = terms
+                    .iter()
+                    .map(|t| matcher.fuzzy(&search_result.header, t, true))
+                    .map(|t| {
+                        if let Some((s, pos)) = t {
+                            (s, pos.first().cloned())
+                        } else {
+                            (0, None)
+                        }
+                    })
+                    .fold((0, Vec::new()), |mut acc, (s, pos)| {
+                        acc.1.push(pos);
+                        (acc.0 + s, acc.1)
+                    });
+
+                header_positions.sort();
+                header_positions.reverse();
+                let header_hit_distance = dt(header_positions.as_slice(), 9999);
+
+                let header_distance_penalty = distance_penalty(dq, header_hit_distance, &terms);
+
+                header_score += 10 * (6 - search_result.ancestry.len()) as i64;
+                header_score -= header_distance_penalty as i64;
+
+                let (mut text_score, mut text_positions) = terms
```

Comment:
Extract duplicate logic

Reply:
have left this out of this PR. would like to tackle this along with a couple other changes to the search step:
- header bonuses at the tantivy level
- change proximity scoring to mimic that of lucene/mellisearch